File: python/infinicore/ops/addbmm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor

# 注意：增加了 out 参数，以及 beta 和 alpha 的默认值
def addbmm(input, batch1, batch2, *, beta=1.0, alpha=1.0, out=None):
    # 1. Out-of-place 模式 (如果没有指定 out)
    if out is None:
        # 错误修复：去掉 .ops，直接调用 _infinicore.addbmm
        # 记得要取 ._underlying 传给 C++
        return Tensor(_infinicore.addbmm(
            input._underlying, 
            batch1._underlying, 
            batch2._underlying, 
            beta, 
            alpha
        ))

    # 2. In-place 模式 (指定了 out)
    # 假设 C++ 端有 addbmm_ (带下划线) 的原地版本，或者重载了支持 out 的版本
    # 这里通常是把 out 作为第一个参数或者通过特定绑定调用
    _infinicore.addbmm_(
        out._underlying, 
        input._underlying, 
        batch1._underlying, 
        batch2._underlying, 
        beta, 
        alpha
    )

    return out

File: src/infinicore/ops/addbmm/addbmm.cc
#include "infinicore/ops/addbmm.hpp"
#include "infinicore/ops/addbmm.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

// 1. 初始化 Dispatcher
common::OpDispatcher<Addbmm::schema> &Addbmm::dispatcher() {
    static common::OpDispatcher<Addbmm::schema> dispatcher_;
    return dispatcher_;
};

// 2. 类层面的 execute 方法
// 【关键修复】参数顺序改为 float beta, float alpha
void Addbmm::execute(Tensor output, Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {
    
    // 切换上下文
    infinicore::context::setDevice(output->device());

    // 分发计算
    dispatcher().lookup(output->device().getType())(output, input, batch1, batch2, beta, alpha);
}

// 3. Out-of-place 接口
// 【关键修复】参数顺序改为 float beta, float alpha
Tensor addbmm(Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    Addbmm::execute(output, input, batch1, batch2, beta, alpha);
    return output;
}

// 4. In-place 接口
// 【关键修复】参数顺序改为 float beta, float alpha
void addbmm_(Tensor output, Tensor input, Tensor batch1, Tensor batch2, float beta, float alpha) {
    Addbmm::execute(output, input, batch1, batch2, beta, alpha);
}

} // namespace infinicore::op